src/index.ts

```diff
@@ -0,0 +1,124 @@
+class School {
+  directions: Direction[] = [];
+
+  addDirection(direction: Direction) {
+    this.directions.push(direction);
+  }
+}
+
+class Direction {
+  levels: Level[] = [];
+  private _name: string;
+
+  get name(): string {
+    return this._name;
+  }
+
+  constructor(name: string) {
+    this._name = name;
+  }
+
+  addLevel(level: Level) {
+    this.levels.push(level);
+  }
+}
+
+class Level {
+  groups: Group[] = [];
+  private _name: string;
+  private _program: string;
+
+  constructor(name: string, program: string) {
+    this._name = name;
+    this._program = program;
+  }
+
+  get name(): string {
+    return this._name;
+  }
+
+  get program(): string {
+    return this._program;
+  }
+
+  addGroup(group: Group): void {
+    this.groups.push(group);
+  }
+}
+
+class Group {
+  private _students: Student[] = [];
+  direction: Direction;
+  level: Level;
+
+  get students(): Student[] {
+    return this._students;
+  }
+
+  constructor(direction: Direction, level: Level) {
+    this.direction = direction;
+    this.level = level;
+  }
+
+  addStudent(student: Student): void {
+    this._students.push(student);
+  }
+
+  showPerformance(): Student[] {
+    const sortedStudents = this.students.sort(
+      (a, b) => b.getPerformanceRating() - a.getPerformanceRating()
+    );
+
+    return sortedStudents;
+  }
+}
+
+class Student {
+  grades: number[] = [];
+  attendance: boolean[] = [];
+  firstName: string;
+  lastName: string;
+  birthYear: number;
+
+  constructor(firstName: string, lastName: string, birthYear: number) {
+    this.firstName = firstName;
+    this.lastName = lastName;
+    this.birthYear = birthYear;
+  }
+
+  get fullName(): string {
+    return `${this.lastName} ${this.firstName}`;
+  }
+
+  set fullName(value: string) {
+    [this.lastName, this.firstName] = value.split(" ");
+  }
+
+  get age(): number {
+    return new Date().getFullYear() - this.birthYear;
+  }
+
+  setGrade(subject: any, grade: number): void {
+    this.grades[subject] = grade;
+  }
+
+  markAttendance(present: boolean): void {
+    this.attendance.push(present);
+  }
+
+  getPerformanceRating(): number {
+    if (this.grades.length === 0) return 0;
+
+    const gradeValues = Object.values(this.grades);
+
+    const averageGrade: number =
+      gradeValues.reduce((sum, grade) => sum + grade, 0) / gradeValues.length;
+
+    const attendancePercentage =
+      (this.attendance.filter((present) => present).length /
```

Comment:
Kind reminder: It is a great opportunity to add explicit type annotations for constants and parameters in the callback function.